### PR TITLE
Warn write-only fields

### PR DESCRIPTION
**Pull request by sanxiyn**
cc @Boscop's example in #49256.

### CODE REVIEW

File: compiler/rustc_passes/src/dead.rs [hunk 1]
```diff
@@ -37,6 +37,19 @@ fn should_explore(tcx: TyCtxt<'_>, hir_id: hir::HirId) -> bool {
...
+    let mut current = expr;
+    loop {
+        match current.kind {
+            hir::ExprKind::Field(base, ..) => {
+                current = base;
+            }
+            _ => break,
+        }
+    }
+    current
```
Comment (oli-obk) [inline compiler/rustc_passes/src/dead.rs:50]: I couldn't resist the code golf opportunity

```suggestion
fn base_expr<'a>(mut expr: &'a hir::Expr<'a>) -> &'a hir::Expr<'a>
    loop {
        match expr.kind {
            hir::ExprKind::Field(base, ..) => expr = base,
            _ => return expr,
        }
    }
}
```

File: compiler/rustc_passes/src/dead.rs [hunk 2]
```diff
@@ -37,6 +37,15 @@ fn should_explore(tcx: TyCtxt<'_>, hir_id: hir::HirId) -> bool {
     )
 }
 
+fn base_expr<'a>(mut expr: &'a hir::Expr<'a>) -> &'a hir::Expr<'a>
```
Comment (oli-obk) [inline compiler/rustc_passes/src/dead.rs:40]: I did a screwup

```suggestion
fn base_expr<'a>(mut expr: &'a hir::Expr<'a>) -> &'a hir::Expr<'a> {
```

**Comment by pnkfelix** [comment]
We discussed this PR in the [compiler triage meeting](https://zulip-archive.rust-lang.org/238009tcompilermeetings/27332weeklymeeting2021031154818.html#229869685).

In short #81473 injected some issues, namely #81626 and #81658, that have leaked into the beta release.
 * #81626 was subsequently resolved on nightly.
 * #81658 is not yet resolved on nightly, but work is on-going to resolve it.

We decided in the compiler triage meeting to revert #81473 on the 1.51 beta branch. We'll leave it in place on nightly 1.52 branch, since we expect #81658 will be resolved before the nightly-to-beta-1.52 promotion.

I'll be posting a revert PR shortly.

**Pull request merged**